### DISCUSSION TITLE
Different Keystore/Alias per partner

### DISCUSSION THREAD

**Discussion by sopgreg** [phax/phase4#117] [Q&A]
Hi,

we are currently using a different AS4 library (using the ENTSOG profile) and are experiencing a problem with phase4 as drop-in replacement:

The other library allows for different keystores/aliases per remote partner. In fact, all private keypairs are stored within one single JKS/PKCS12 file and the approriate alias is stored within the PMode per partner.

The same goes for the encryption keys (certifactes), which are all stored in one JKS/PKCS12 file and the appropriate alias per partner ist also stored in the PMode.

That's how we also built our GUI and storage around, which seemed fine so far.

With phase4, it seems there is only one global keystore set via `org.apache.wss4j.crypto.merlin.keystore(.file|alias|password)`.

Is it possible to change this to a multi-alias behaviour? I tried something like this:

```
Phase4ENTSOGSender.builder()
   .cryptoFactory(xxx)
``` 

But I would actually only need to override the `getKeyAlias` depending on the partner. And what about the reception side for decryption?

Another approach, which would also be fine, could be to store a separate PKCS12 file per remote partner.

**Comment by phax** [comment]
Hi @sopgreg,

Please see https://github.com/phax/phase4/discussions/88 for a very similar discussion.
If you have further questions, don't hesitate to ask.

hth, Philip

**Reply by sopgreg** [reply]
Hi @phax,

I've seen this discussion. But what I'm wondering is, how I would specify a specific `KeyStore` or `PrivateKeyEntry` (as defined in `IAS4CryptoFactory`) based on the PMode used. There seems to be missing some context from within this callback such as the initiator (for outgoing files, and vice versa for incoming files).

**Reply by phax** [reply]
Well, basically you can make the decision on whatever parameter you like - at least for sending.
Once you made the decision, you create a new `AS4CryptoFactoryInMemoryKeyStore` and add it via `.cryptoFactory (...)` to the sending builder.

For reception it becomes a bit more tricky. Usually the determination needs to be done on a URL level or on something else. Finding the certificate based on the SOAP header is currently not foreseen :/

**Reply by sopgreg** [reply]
Ok thanks, the sending side is simple via `.cryptoFactory` in the builder and `AS4CryptoFactoryInMemoryKeyStore`, I managed to get this to work.

What would be an appropriate code location to look into for the receiving side in phase4? Maybe there's a place where I can hack around a bit and see what works.

**Reply by phax** [reply]
Unfortunately there isn't really. The basic way to define a custom cryptoFactory on receiving side is described in https://github.com/phax/phase4/discussions/88#discussioncomment-2956501 - create your own servlet and customize the handler accordingly.

Do make that dependent on the PMode that means passing some new interface all the way in. I believe that to  be tricky

**Reply by sopgreg** [reply]
I could imagine a small link here, would that be a feasable solution worth a try? 

`SOAPHeaderElementProcessorRegistry:92`

```
    final SOAPHeaderElementProcessorRegistry ret = new SOAPHeaderElementProcessorRegistry ();
    ret.registerHeaderElementProcessor (SOAPHeaderElementProcessorExtractEbms3Messaging.QNAME_MESSAGING,
                                        new SOAPHeaderElementProcessorExtractEbms3Messaging (aPModeResolver));

    // WSS4J must be after Ebms3Messaging handler!
    ret.registerHeaderElementProcessor (SOAPHeaderElementProcessorWSS4J.QNAME_SECURITY,
                                        new SOAPHeaderElementProcessorWSS4J (aCryptoFactory, aFallbackPMode));
```

Introduce a small callback in `SOAPHeaderElementProcessorExtractEbms3Messaging`, once the PMode was successfully resolved. Something like

```
public interface IPModeResolvedCallback {
   void onPModeResolved(IPMode pMode) {};
}
```

The listener of this callback (local to this method) would inform the `aCryptoFactory` about the PMode in context, such like:

```
new SOAPHeaderElementProcessorExtractEbms3Messaging (aPModeResolver, pModeResolved -> aCryptoFactory.setContextPMode(pModeResolved));
```

This would require a new (default) method in the `IAS4CryptoFactory` such like:

```
public interface IAS4CryptoFactory {
   default void setContextPMode(IPMode pMode) {};
}
```

Custom classes could then create a simple `IAS4CryptoFactory` implementation that can lookup the key stores based on the PMode in the current context.

**Reply by phax** [reply]
Without being in front of the code I think that looks feasible. The only thing I would change is to introduce a new interface `IPModeAwareCertificateFactory` with the new method, to keep backwards compatibility. That may lead to active CryptoFactory implementations nit being `final` - but that's a micro thingy...

Also I like the default crypto factories to be read-only for the sake of simplicity and thread management etc.

**Reply by sopgreg** [reply]
That was also something I was worried about, a new `IPModeAwareCertificateFactory` sounds fine for me, too. It would just require an `instanceof` check in the `IPModeResolvedCallback`, like (pseudocode):

```
new SOAPHeaderElementProcessorExtractEbms3Messaging (aPModeResolver, 
   pModeResolved -> if (aCryptoFactory instanceof IPModeAwareCertificateFactory cf) { cf.setContextPMode(pModeResolved)); }
```

Is the whole feature something you would consider adding to the main branch anytime soon, or would you expect a pull request?

**Reply by phax** [reply]
If you find time for a PR I would appreciate it - and I don't expect it to be perfect, but if it works for you, we can save the testing iteration :)

**Reply by sopgreg** [reply]
I created a PR with the changes which did work for us

**Reply by phax** [reply]
So PR #121  was merged and will be part of the 2.1.0 release
